// @shapescape/storage/dist/index.mjs
// src/storage/property-storage.ts
var PropertyStorage = class _PropertyStorage {
  /**
   * Maximum allowed byte size for a value.
   * @private
   */
  MAX_BYTE_SIZE = 32 * 1024 - 1;
  /**
   * The underlying storage backend.
   */
  storage;
  /**
   * Optional prefix for keys.
   */
  prefix;
  /**
   * Creates a new PropertyStorage instance.
   * @param storage - The dynamic storage object.
   * @param prefix - Optional key prefix.
   */
  constructor(storage, prefix) {
    this.storage = storage;
    this.prefix = prefix ?? "";
  }
  /**
   * Sets a value in the storage.
   * @param key - The key to set.
   * @param value - The value to set, can be any JSON-serializable value.
   * @param options - Optional serialization options.
   */
  set(key, value, options) {
    key = this.prefix + key;
    value = this.serialize(value, options);
    this.getStorage().setDynamicProperty(key, value);
  }
  /**
   * Removes a value from the storage.
   * @param key - The key to remove.
   */
  drop(key) {
    key = this.prefix + key;
    this.getStorage().setDynamicProperty(key);
  }
  /**
   * Clears all values from the storage.
   * This method deletes keys matching the pattern in the main storage.
   * If includeSubStorages is true, it also recursively clears values within each sub-storage,
   * leaving the sub-storage container itself intact.
   * @param pattern - Pattern to match keys against. If omitted, defaults to the storage prefix.
   * @param includeSubStorages - Whether to clear values in sub-storages.
   */
  clear(pattern, includeSubStorages = true) {
    const effectivePattern = pattern ? this.prefix + pattern : this.prefix;
    const keys = this.getStorage().getDynamicPropertyIds();
    const subStorages = this.get("subStorages", false, []);
    for (const key of keys) {
      if (key.startsWith(effectivePattern)) {
        if (subStorages.includes(key)) continue;
        this.getStorage().setDynamicProperty(key);
      }
    }
    if (includeSubStorages) {
      for (const sub of subStorages) {
        const subStore = this.getSubStorage(sub);
        subStore.clear(void 0, false);
      }
    }
  }
  /**
   * Gets a value from the storage.
   * @param key - The key to get.
   * @param deserialize - Whether to deserialize the value.
   * @param defaultValue - A default value returned if the key does not exist.
   * @param options - Optional deserialization options.
   * @returns The stored value or the default value.
   */
  get(key, deserialize = true, defaultValue, options) {
    key = this.prefix + key;
    const value = this.getStorage().getDynamicProperty(key);
    if (value === void 0) return defaultValue;
    if (typeof value === "string" && deserialize)
      return this.deserialize(value, options);
    return value;
  }
  /**
   * Gets all values from the storage.
   * @param pattern - Pattern to match keys against.
   * @param deserialize - Whether to deserialize the returned values.
   * @param options - Optional deserialization options.
   * @returns Array of key-value pair objects.
   */
  getAll(pattern, deserialize, options) {
    const properties = [];
    const keys = this.getStorage().getDynamicPropertyIds();
    if (!pattern) pattern = this.prefix;
    else pattern = this.prefix + pattern;
    for (let key of keys) {
      if (key.startsWith(pattern)) {
        key = key.slice(this.prefix.length);
        let value = this.get(key, deserialize, void 0, options);
        properties.push({ [key]: value });
      }
    }
    return properties;
  }
  /**
   * Pushes a value to the end of an array stored in the storage.
   * @param key - The key of the array.
   * @param value - The value to push.
   */
  rPush(key, value) {
    const property = this.assertArray(key, true);
    property.push(value);
    this.set(key, property);
  }
  /**
   * Pushes a value to the beginning of an array stored in the storage.
   * @param key - The key of the array.
   * @param value - The value to push.
   */
  lPush(key, value) {
    const property = this.assertArray(key, true);
    property.unshift(value);
    this.set(key, property);
  }
  /**
   * Pops a value from the end of an array.
   * @param key - The key of the array.
   * @returns The value popped.
   */
  rPop(key) {
    const property = this.assertArray(key);
    const value = property.pop();
    this.set(key, property);
    return value;
  }
  /**
   * Pops a value from the beginning of an array.
   * @param key - The key of the array.
   * @returns The value popped.
   */
  lPop(key) {
    const property = this.assertArray(key);
    const value = property.shift();
    this.set(key, property);
    return value;
  }
  /**
   * Calculates the byte size of a stored value.
   * @param key - The key of the value.
   * @returns The byte length.
   */
  getByteSize(key) {
    return this.encodeValue(this.get(key, false)).byteLength;
  }
  /**
   * Retrieves the underlying storage object.
   * @returns The dynamic storage.
   */
  getStorage() {
    return this.storage;
  }
  /**
   * Builds a sub-storage prefix based on a given identifier.
   * @param prefix - The sub-storage identifier.
   * @returns The full prefix.
   */
  getSubStoragePrefix(prefix) {
    return this.prefix + "_" + prefix + "_";
  }
  /**
   * Registers a sub-storage.
   * @param prefix - The sub-storage identifier.
   */
  registerSubStorage(prefix) {
    const subStorages = this.get("subStorages", true, []);
    if (!subStorages.includes(prefix)) {
      subStorages.push(prefix);
      this.set("subStorages", subStorages);
    }
  }
  /**
   * Retrieves a sub-storage.
   * @param prefix - The sub-storage identifier.
   * @returns A new PropertyStorage instance with the updated prefix.
   */
  getSubStorage(prefix) {
    this.registerSubStorage(prefix);
    return new _PropertyStorage(this.storage, this.getSubStoragePrefix(prefix));
  }
  /**
   * Encodes a value to a Uint8Array.
   * @param value - The value to encode.
   * @returns The encoded Uint8Array.
   */
  encodeValue(value) {
    const binStr = decodeURIComponent(encodeURIComponent(value)), arr = new Uint8Array(binStr.length);
    const split = binStr.split("");
    for (let i = 0; i < split.length; i++) {
      arr[i] = split[i].charCodeAt(0);
    }
    return arr;
  }
  /**
   * Serializes a value to a JSON string.
   * @param data - The data to serialize.
   * @param options - Optional serialization options.
   * @returns The serialized string.
   */
  serialize(data, options) {
    if (typeof data !== "string")
      return JSON.stringify(data, options?.replacer, options?.space);
    return data;
  }
  /**
   * Deserializes a JSON string to a value.
   * @param data - The string to deserialize.
   * @param options - Optional deserialization options.
   * @returns The deserialized value.
   */
  deserialize(data, options) {
    try {
      return JSON.parse(data, options?.reviver);
    } catch (ignored) {
      return data;
    }
  }
  /**
   * Ensures that a storage key corresponds to an array.
   * @param key - The key to verify.
   * @param init - Whether to initialize the array if not exists.
   * @returns The array stored at the key.
   */
  assertArray(key, init = false) {
    let property = this.get(key);
    if (!property) {
      property = [];
      if (init) this.set(key, []);
    } else if (!Array.isArray(property)) {
      throw new TypeError(`${key} is not an array`);
    }
    return property;
  }
  /**
   * Stores multiple key-value pairs.
   * @param pairs - An object with keys and corresponding values.
   * @param options - Optional serialization options.
   */
  storeBulk(pairs, options) {
    for (const key in pairs) {
      this.set(key, pairs[key], options);
    }
  }
  /**
   * Retrieves multiple keys at once.
   * @param keys - An array of keys.
   * @param deserialize - Whether to deserialize the values.
   * @param options - Optional deserialization options.
   * @returns An array of values.
   */
  fetchBulk(keys, deserialize = true, options) {
    return keys.map((key) => this.get(key, deserialize, void 0, options));
  }
  /**
   * Checks whether a key exists.
   * @param key - The key to check.
   * @returns True if exists, false otherwise.
   */
  keyExists(key) {
    return this.get(key) !== void 0;
  }
  /**
   * Searches keys matching a given pattern.
   * @param pattern - The pattern to match.
   * @returns An array of matching keys (without the prefix).
   */
  searchKeys(pattern) {
    const allKeys = this.getStorage().getDynamicPropertyIds();
    if (!pattern) return allKeys.map((k) => k.substring(this.prefix.length));
    return allKeys.filter((k) => k.startsWith(this.prefix + pattern)).map((k) => k.substring(this.prefix.length));
  }
};

// src/storage/multi-array.ts
var MultiArray = class _MultiArray {
  /**
   * The key identifying the array.
   */
  key;
  /**
   * The ExtendedStorage instance.
   */
  storage;
  /**
   * The current array index.
   */
  index;
  /**
   * The sub-index information for navigation.
   */
  subIndex;
  /**
   * Creates a new MultiArray instance.
   * @param key - The array key.
   * @param storage - The ExtendedStorage instance.
   * @param index - The index of the current element.
   * @param subIndex - The sub-index containing previous and next pointers.
   */
  constructor(key, storage, index, subIndex) {
    this.key = key;
    this.storage = storage;
    this.index = index;
    this.subIndex = subIndex;
  }
  /**
   * Retrieves the value of the current element.
   * @returns The element's value, or undefined if removed.
   */
  getValue() {
    if (this.isRemoved()) return void 0;
    return this.storage.get(this.key + ".ar." + this.index);
  }
  /**
   * Sets the value of the current element.
   * @param value - The value to set.
   * @throws Error if the element has been removed.
   */
  setValue(value) {
    if (this.isRemoved())
      throw new Error("Cannot set value of removed MultiArray element.");
    this.storage.set(this.key + ".ar." + this.index, value);
  }
  /**
   * Retrieves the next element in the array.
   * @returns A new MultiArray instance for the next element or null if at the end.
   */
  getNext() {
    if (this.subIndex.next === null) return null;
    const subIndex = this.storage.getSubIndex(this.key, this.subIndex.next);
    if (!subIndex) return null;
    return new _MultiArray(this.key, this.storage, this.subIndex.next, subIndex);
  }
  /**
   * Retrieves the previous element in the array.
   * @returns A new MultiArray instance for the previous element or null if at the beginning.
   */
  getPrev() {
    if (this.subIndex.prev === null) return null;
    const subIndex = this.storage.getSubIndex(this.key, this.subIndex.prev);
    if (!subIndex) return null;
    return new _MultiArray(this.key, this.storage, this.subIndex.prev, subIndex);
  }
  /**
   * Checks whether the current element has been removed.
   * @returns True if removed.
   */
  isRemoved() {
    return this.storage.get(this.key + ".ar." + this.index) === void 0;
  }
  /**
   * Removes the current element from the array and updates neighbors.
   */
  remove() {
    if (this.subIndex.prev !== null) {
      const prev = this.storage.get(this.key + ".ar." + this.subIndex.prev);
      prev.next = this.subIndex.next;
      this.storage.set(this.key + ".ar." + this.subIndex.prev, prev);
    }
    if (this.subIndex.next !== null) {
      const next = this.storage.get(this.key + ".ar." + this.subIndex.next);
      next.prev = this.subIndex.prev;
      this.storage.set(this.key + ".ar." + this.subIndex.next, next);
    }
    this.storage.drop(this.key + ".ar." + this.index);
  }
  /**
   * Inserts a new element after the current element.
   * @param value - The value to insert.
   */
  insertAfter(value) {
    const index = this.storage.getArrayIndex(this.key);
    if (!index) throw new Error("Array index not found.");
    const newIndex = index.end + 1;
    const newSubIndex = {
      prev: this.index,
      next: this.subIndex.next
    };
    this.subIndex.next = newIndex;
    this.storage.set(this.key + ".ar." + this.index, this.subIndex);
    this.storage.set(this.key + ".ar." + newIndex, newSubIndex);
    this.storage.set(this.key + ".ar." + newIndex, value);
  }
  /**
   * Inserts a new element before the current element.
   * @param value - The value to insert.
   */
  insertBefore(value) {
    const index = this.storage.getArrayIndex(this.key);
    if (!index) throw new Error("Array index not found.");
    const newIndex = index.start - 1;
    const newSubIndex = {
      prev: this.subIndex.prev,
      next: this.index
    };
    this.subIndex.prev = newIndex;
    this.storage.set(this.key + ".ar." + this.index, this.subIndex);
    this.storage.set(this.key + ".ar." + newIndex, newSubIndex);
    this.storage.set(this.key + ".ar." + newIndex, value);
  }
};

// src/storage/extended-storage.ts
var ExtendedStorage = class _ExtendedStorage extends PropertyStorage {
  /**
   * Initialization flag.
   * @private
   */
  init = false;
  /**
   * Checks whether the provided storage is an ExtendedStorage.
   * @param storage - The storage instance.
   * @returns True if extended.
   */
  static isExtendedStorage(storage) {
    return storage.get("shape_extended_storage");
  }
  /**
   * Initializes the extended storage.
   * @private
   */
  initStorage() {
    super.set("shape_extended_storage", true);
    this.init = true;
  }
  /**
   * Sets the index for a value.
   * @param key - The key of the value.
   * @param index - The index definition.
   */
  setIndex(key, index) {
    super.set(key + ".index", index);
    if (!this.init) this.initStorage();
  }
  /**
   * Gets the index for a value.
   * @param key - The key of the value.
   * @returns The index definition.
   */
  getIndex(key) {
    const value = super.get(key + ".index");
    if (value) {
      if (!value.max) throw new TypeError(`Key ${key} is not an index`);
      return value;
    }
    return void 0;
  }
  /**
   * Sets the array index structure.
   * @param key - The key of the array.
   * @param index - The array index definition.
   */
  setArrayIndex(key, index) {
    super.set(key, index);
    if (!this.init) this.initStorage();
  }
  /**
   * Gets the array index structure.
   * @param key - The key of the array.
   * @returns The array index definition.
   */
  getArrayIndex(key) {
    const value = super.get(key + ".index");
    if (value) {
      if (!value.start) throw new TypeError(`Key ${key} is not an array index`);
      return value;
    }
    return void 0;
  }
  /**
   * Sets a value in extended storage. For values larger than 32KiB, splits into chunks.
   * @param key - The key to set.
   * @param value - The value to store.
   * @param options - Optional serialization options.
   */
  set(key, value, options) {
    let index = this.getIndex(key);
    if (index || !value) {
      this.drop(key);
    }
    if (!value) return;
    if (/\.\d+$/.test(key)) {
      return super.set(key, value);
    }
    const serializedValue = this.serialize(value, options);
    const chunks = this.splitValueIntoChunks(serializedValue);
    index = { max: chunks.length };
    this.setIndex(key, index);
    for (let i = 0; i < chunks.length; i++) {
      super.set(key + "." + i, chunks[i]);
    }
  }
  /**
   * Gets a value from extended storage.
   * @param key - The key to retrieve.
   * @param deserialize - Whether to deserialize the string result.
   * @param defaultValue - The default value if not found.
   * @param options - Optional deserialization options.
   * @returns The stored value.
   */
  get(key, deserialize = true, defaultValue, options) {
    if (/\.\d+$/.test(key)) {
      return super.get(key, deserialize, defaultValue, options);
    }
    const index = this.getIndex(key);
    if (!index) return defaultValue;
    let result = "";
    for (let i = 0; i < index.max; i++) {
      const chunk = super.get(key + "." + i, false);
      if (!chunk) return defaultValue;
      result += chunk;
    }
    if (result === "") return defaultValue;
    if (deserialize) {
      return this.deserialize(result, options);
    }
    return result;
  }
  /**
   * Removes a value from extended storage.
   * @param key - The key to remove.
   */
  drop(key) {
    if (this.isArray(key)) {
      const arrayIndex = this.getArrayIndex(key);
      if (!arrayIndex) return;
      let i = arrayIndex.start;
      while (i != null) {
        super.drop(key + ".ar." + i);
        i = arrayIndex[i].next;
      }
      this.setArrayIndex(key, void 0);
    } else {
      const index = this.getIndex(key);
      if (!index) {
        super.drop(key);
        return;
      }
      for (let i = 0; i < index.max; i++) {
        super.drop(key + "." + i);
      }
      this.setIndex(key, void 0);
    }
  }
  /**
   * Retrieves a MultiArray instance to navigate a linked list array.
   * @param key - The key of the array.
   * @returns The MultiArray instance.
   */
  getMultiArray(key) {
    const arrayIndex = this.getArrayIndex(key);
    if (!arrayIndex) throw new TypeError(`Key ${key} is not an array`);
    if (this.isArrayEmpty(key))
      throw new TypeError(`Key ${key} is an empty array`);
    const subIndex = arrayIndex[arrayIndex.start];
    return new MultiArray(key, this, arrayIndex.start, subIndex);
  }
  /**
   * Appends a value to the end of an array.
   * @param key - The array key.
   * @param value - The value to append.
   * @param options - Optional serialization options.
   */
  rPush(key, value, options) {
    const index = this.assertMultiArray(key, true);
    if (index.end === index.start) {
      index[index.end] = { prev: null, next: null };
    } else {
      index[index.end].next = index.end + 1;
      index[index.end + 1] = { prev: index.end, next: null };
    }
    index.end++;
    super.set(key + ".ar." + index.end, value, options);
    this.setArrayIndex(key, index);
  }
  /**
   * Prepends a value to the start of an array.
   * @param key - The array key.
   * @param value - The value to prepend.
   * @param options - Optional serialization options.
   */
  lPush(key, value, options) {
    const index = this.assertMultiArray(key, true);
    if (index.end === index.start) {
      index[index.start] = { prev: null, next: null };
    } else {
      index[index.start].prev = index.start - 1;
      index[index.start - 1] = { prev: null, next: index.start };
    }
    index.start--;
    super.set(key + ".ar." + index.start, value, options);
    this.setArrayIndex(key, index);
  }
  /**
   * Removes and returns the last element of an array.
   * @param key - The array key.
   * @param deserialize - Whether to deserialize the value.
   * @param options - Optional deserialization options.
   * @returns The removed element.
   */
  rPop(key, deserialize, options) {
    const index = this.assertMultiArray(key);
    if (index.end === index.start) return void 0;
    const value = super.get(key + ".ar." + index.end, deserialize, options);
    index.end--;
    this.setArrayIndex(key, index);
    return value;
  }
  /**
   * Removes and returns the first element of an array.
   * @param key - The array key.
   * @param deserialize - Whether to deserialize the value.
   * @param options - Optional deserialization options.
   * @returns The removed element.
   */
  lPop(key, deserialize, options) {
    const index = this.assertMultiArray(key);
    if (index.end === index.start) return void 0;
    const value = super.get(key + ".ar." + index.start, deserialize, options);
    index.start++;
    this.setArrayIndex(key, index);
    return value;
  }
  /**
   * Checks whether an array is empty.
   * @param key - The array key.
   * @returns True if empty.
   */
  isArrayEmpty(key) {
    const index = this.getArrayIndex(key);
    return !index || index.end === index.start;
  }
  /**
   * Checks whether the value at a key is an array.
   * @param key - The key to check.
   * @returns True if an array.
   */
  isArray(key) {
    const index = this.getArrayIndex(key);
    return !!index && !this.isArrayEmpty(key);
  }
  /**
   * Retrieves the length of an array.
   * @param key - The array key.
   * @returns The length.
   */
  getArrayLength(key) {
    if (!this.isArray(key)) return 0;
    const index = this.getArrayIndex(key);
    if (!index) return 0;
    let multiArray = this.getMultiArray(key);
    let length = 0;
    while (multiArray) {
      length++;
      multiArray = multiArray.getNext();
    }
    return length;
  }
  /**
   * Gets the sub-index for an element in an array.
   * @param key - The array key.
   * @param index - The element index.
   * @returns The SubIndex definition.
   */
  getSubIndex(key, index) {
    const arrayIndex = this.getArrayIndex(key);
    if (!arrayIndex) return void 0;
    return arrayIndex[index];
  }
  /**
   * Asserts the key is an array and returns its index structure.
   * @param key - The key to assert.
   * @param create - Whether to create the index if it does not exist.
   * @returns The ArrayIndex structure.
   */
  assertMultiArray(key, create = false) {
    let index = this.getArrayIndex(key);
    if (!index) {
      index = { start: 0, end: 0 };
    }
    if (index.end === index.start) {
      index.start = 0;
      index.end = 0;
    }
    if (create) {
      this.setArrayIndex(key, index);
    }
    return index;
  }
  /**
   * Splits a value into chunks of a specified size.
   * @param input - The value to split.
   * @param chunkSize - The maximum size per chunk (default 32KiB).
   * @returns An array of chunk strings.
   */
  splitValueIntoChunks(input, chunkSize = this.MAX_BYTE_SIZE) {
    const encodedInput = this.encodeValue(input);
    const result = [];
    let offset = 0;
    while (offset < encodedInput.length) {
      const size = Math.min(chunkSize, encodedInput.length - offset);
      const chunk = encodedInput.subarray(offset, offset + size);
      const decoder = new TextDecoder();
      result.push(decoder.decode(chunk));
      offset += size;
    }
    return result;
  }
  /**
   * Retrieves a sub-storage with a given prefix.
   * @param prefix - The sub-storage identifier.
   * @returns A new ExtendedStorage instance with the sub-storage prefix.
   */
  getSubStorage(prefix) {
    this.registerSubStorage(prefix);
    return new _ExtendedStorage(
      this.getStorage(),
      this.getSubStoragePrefix(prefix)
    );
  }
};

// src/storage/cached-storage.ts
var CachedStorage = class _CachedStorage extends PropertyStorage {
  /**
   * In-memory cache storage.
   */
  cache = /* @__PURE__ */ new Map();
  /**
   * Sets a value in the storage and caches it.
   * @param key - The key to set.
   * @param value - The value (must be JSON-serializable).
   * @param options - Optional serialization options.
   */
  set(key, value, options) {
    key = this.prefix + key;
    value = this.serialize(value, options);
    this.getStorage().setDynamicProperty(key, value);
    this.cache.set(key, value);
  }
  /**
   * Removes a value from both the storage and the in-memory cache.
   * @param key - The key to remove.
   */
  remove(key) {
    key = this.prefix + key;
    this.getStorage().setDynamicProperty(key);
    this.cache.delete(key);
  }
  /**
   * Clears values from the storage and cache.
   * This method deletes keys matching the pattern in the underlying storage by calling super.clear(),
   * based on the includeSubStorages flag, and then removes matching entries from the in-memory cache.
   * @param pattern - Pattern to match keys against. If omitted, defaults to the storage prefix.
   * @param includeSubStorages - Whether to clear values in sub-storages.
   */
  clear(pattern, includeSubStorages = true) {
    super.clear(pattern, includeSubStorages);
    const effectivePattern = pattern ? this.prefix + pattern : this.prefix;
    for (const key of this.cache.keys()) {
      if (key.startsWith(effectivePattern)) {
        this.cache.delete(key);
      }
    }
  }
  /**
   * Retrieves a value from the cache or storage.
   * @param key - The key to fetch.
   * @param deserialize - Whether to deserialize the value.
   * @param defaultValue - The default value if the key does not exist.
   * @param options - Optional deserialization options.
   * @returns The fetched value.
   */
  get(key, deserialize = true, defaultValue, options) {
    key = this.prefix + key;
    let value = this.cache.get(key);
    if (value === void 0) value = this.getStorage().getDynamicProperty(key);
    if (value === void 0) return defaultValue;
    if (typeof value === "string" && deserialize)
      return this.deserialize(value, options);
    return value;
  }
  /**
   * Retrieves a sub-storage instance with a given prefix.
   * @param prefix - The sub-storage identifier.
   * @returns A new CachedStorage instance.
   */
  getSubStorage(prefix) {
    this.registerSubStorage(prefix);
    return new _CachedStorage(
      this.getStorage(),
      this.getSubStoragePrefix(prefix)
    );
  }
};
export {
  CachedStorage,
  ExtendedStorage,
  MultiArray,
  PropertyStorage
};
